wip-trampoline.md

```diff
@@ -0,0 +1,780 @@
+* [`80`:`hop_payload`] (payload for T)
+  * [`1`:`0x4e`] (`length`)
+  * [`1`:`0x02`] (`type`)
+  * [`1`:`0x03`] (`length`)
+  * [`3`:`5000000`] (`amt_to_forward`)
+  * [`1`:`0x04`] (`type`)
+  * [`1`:`0x01`] (`length`)
+  * [`1`:`25`] (`outoing_cltv_value`)
+  * [`1`:`0x0a`] (`type`)
+  * [`1`:`0x21`] (`length`)
+  * [`33`:`Bob_node_id`]
+  * [`1`:`0x0c`] (`type`)
+  * [`1`:`0x02`] (`length`)
+  * [`2`:`0x0000`] (`trampoline_routing` feature flags disabled)
+  * [`32`:`hmac`] (`0x00...00`)
+* [`320`:`filler`]
+* [`32`:`hmac`]
+
+Alice finds a route to T and sends the `trampoline_onion_packet` wrapped inside an `onion_packet` (see previous
+example).
+
+T receives the `trampoline_onion_packet` and discovers that Bob is the payment recipient and doesn't support the
+`trampoline_routing` feature.
+
+T finds a route to Bob:
+
+```text
+T -> H1 -> H2 -> Bob
+```
+
+T creates the following `onion_packet` (encryption omitted for clarity):
+
+* [`1`:`0x00`] (`version`)
+* [`33`:`public_key`]
+* [`65`:`hop_payload`] (payload for H1)
+  * [`1`:`0x00`] (`realm`)
+  * [`8`:`channel_from_H1_to_H2`] (`short_channel_id`)
+  * [`8`:`5000500`] (`amt_to_forward`)
+  * [`4`:`30`] (`outgoing_cltv_value`)
+  * [`12`:`padding`]
+  * [`32`:`hmac`]
+* [`65`:`hop_payload`] (payload for H2)
+  * [`1`:`0x00`] (`realm`)
+  * [`8`:`channel_from_H2_to_Bob`] (`short_channel_id`)
+  * [`8`:`5000000`] (`amt_to_forward`)
+  * [`4`:`25`] (`outgoing_cltv_value`)
+  * [`12`:`padding`]
+  * [`32`:`hmac`]
+* [`65`:`hop_payload`] (payload for Bob)
+  * [`1`:`0x00`] (`realm`)
+  * [`8`:`0x0000000000000000`] (`short_channel_id`)
+  * [`8`:`5000000`] (`payment_amt`)
+  * [`4`:`25`] (`final_cltv_expiry`)
+  * [`12`:`padding`]
+  * [`32`:`hmac`] (`0x00...00`)
+* [`1105`:`filler`]
+* [`32`:`hmac`]
+
+Bob receives the `onion_packet` and discovers that he is the recipient of the payment (because the hmac is `0x00...00`).
+Bob can process the payment and doesn't know trampoline routing was used.
+
+The effective payment route is:
+
+```text
+Alice      T                   Bob
+  |       ^ |                   ^
+  |       | |                   |
+  `-------' `---> H1 ---> H2 ---'
+```
+
+## Appendix B: Trampoline and Rendez-Vous
+
+A mechanism to implement rendez-vous routing was previously proposed [here][rendez-vous].
+It uses a new mechanism that allows switching the `session_key` in the middle of the route.
+
+Trampoline routing provides another option to implement rendez-vous routing (with some similarities).
+The recipient creates a small onion from his rendez-vous point to himself and embeds it inside a payment request.
+The sender includes that small onion in his normal onion payload and it gets relayed to the last trampoline node.
+That last trampoline node then uses the provided rendez-vous onion to reach the recipient.
+
+## Appendix C: Trampoline and MPP
+
+Trampoline routing combines nicely with multi-part payments. When multi-part payment is used, we can allow trampoline
```

Comment:
Hmmm if a trampoline node has a good view of a network, knowing every last partial payment node, it should be able to do path intersections from them to discover who is the original sender. IMO, a randomly-generated one-shot tunnel with a minimal length would be better than splitting payment if we want to preserve payer-privacy towards first trampoline node. But this whole point need real analysis.

Reply:
Interesting point, thanks for raising this! This is an issue with normal AMP, isn't it? The recipient could de-anonymize the payer by doing path intersection (even though in reality, the payer's channels are likely unannounced channels so the recipient can't de-anonymize the first hop).

If I understand your attack correctly, it seems that on the contrary trampoline helps anonymity in that case. Because in that case a node is only able to de-anonymize the previous node, thus discovering who the previous trampoline hop was. But thanks to onion encryption, this doesn't tell the attacker whether that node was the payer or an intermediate trampoline node! Unless the payer doesn't advertise trampoline feature bits, in which case we are in the same spot as normal AMP privacy-wise.

Reply:
Let's say you have following topology:

```    
        ______  B __________
     /                      \
   /                         \
 A  ___________ C  _________ T
   \                        /
     \ _______ D  _________/

```

If all channels are publicly knew and that's the whole graph, and A is using T as trampoline node by sending a AMP payment through B, C and D, that's super easy to deduct than A is payer from T viewpoint. Now this topology is super dumb, and private channels would obfuscate real graph and make whole analysis probabilistic-based. So payer should use only private channels as first links to broadcast AMP payment. Like you said it should keep you safe but that's easy to mess up in practice ("Oh my two first channels are private, that should be enough, don't care than my third one is public !")

Reply:
But the example you're sharing showcases an issue with AMP _without_ trampoline ;)
If this is a non-trampoline AMP where T is the recipient, T is able to deduce who the payer is.

In the trampoline case, T is only able to deduce that A was the previous trampoline node (which may be the payer of just another intermediate trampoline node). So it's just like knowing the previous hop in a normal payment route.

Reply:
I'm dumb, you're right I forgot a vertice from T the trampoline to some E node the payee.

Yes I think you're right that T is only able to deduce that A was the previous trampoline node (not necessarily the payer), but to keep analysis blurred we should recommend to use multiple trampoline/or private channels. If it's knew that a particular implementation uses only one-hop trampoline privacy would be broken.

Reply:
I completely agree. Implementations should never hard-code or bias too much towards a one-hop trampoline route. I'll make stress this more in the rationale.